process_sandbox documentation & build (#524)
microsoft/verona: experiments/process_sandbox/docs/walkthrough.md

                                                 |        PageMap                               |                                              | 
  |-- 0x7ffe80000000 -- 0x7ffec0000000 [rw-s]    |                                              |                                              |-- child 3 region [rw-s]
                                                 |                                              |                                            
  |-- 0x7fff00000000 -- 0x7fff40000000 [rw-s]    |                                              |-- child 2 region [rw-s]                     
                                                 |                                                                                           
  |-- 0x7fff80000000 -- 0x7fffc0000000 [rw-s]    |-- child 1 region [rw-s]                                                                   

```

The `library_runner` is responsible, at initialization time, for ensuring that both the `PageMap` and the child's memory region (i.e., heap) are correctly mapped in the process's address space.

> For that purposes, it uses the `memfd` file descriptors inherited from the parent.

Reviewer: ```suggestion
For that purposes, it uses the file descriptors referring to shared memory objects that it inherits from the parent.
```

It relies on `mmap` to install the heap as a shared memory region that is both readable and writable.

> The `library_runner` opens the `memfd` file descriptor that corresponds to the `PageMap` object and maps its content read-only in the child's address space.

Reviewer: ```suggestion
The `library_runner` opens the file descriptor that corresponds to the shared memory object that backs the `PageMap` and maps its content read-only in the child's address space.
The platform-specific sandboxing code is responsible for ensuring that the child process cannot perform a read-write mapping of this file descriptor.
```

#### Invoking the sandbox

We follow the execution of the program and show how, on Linux, the parent invokes the `sum` function defined in [sandboxlib-basic](../tests/sandboxlib-basic.cc).

> On the parent side, the sandbox library sets up a call frame containing the arguments to the function call and a return element in the shared memory region and points the `msg_buffer` field in the header of the shared memory region to this buffer.

Reviewer: ```suggestion
On the parent side, the sandbox library creates an object representing a call frame, containing the arguments to the function call and space for any return element, in the shared memory region and points the `msg_buffer` field in the header of the shared memory region to this object.
```

> It passes the pointer to that call frame as well as the index of the library's function invoked via shared memory to the child that it wakes up.

Reviewer: ```suggestion
It also stores the index of the library's function to be invoked in the `function_index` field.
```

Simple one-bit in-memory semaphores are used to communicate with the child process.
These are implemented with the platform's low-level primitive, `futex` on Linux and `_umtx_op` on FreeBSD.
The trace below shows the parent's stack trace:

```
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x00007ffff7f3b611 in sandbox::platform::FutexOneBitSem::futex_op (this=0x7fff800000c0, futex_op=1, val=1, timeout=0x0) at ../include/process_sandbox/platform/onebitsem_futex.h:29
#2  0x00007ffff7f3b884 in sandbox::platform::FutexOneBitSem::wake (this=0x7fff800000c0) at ../include/process_sandbox/platform/onebitsem_futex.h:46
#3  0x00007ffff7f38d77 in sandbox::Library::send (this=0x7fffffffd1b0, idx=1, ptr=0x7fff80004020) at ../src/libsandbox.cc:774
#4  0x000055555557e0b0 in sandbox::Function<int, int, int>::operator() (this=0x7fffffffd660, args#0=1, args#1=2) at ../include/process_sandbox/cxxsandbox.h:103
#5  0x000055555557cd65 in test_sum (sb=..., a=1, b=2) at ../tests/sandbox-basic.cc:31
#6  0x000055555557cf18 in main () at ../tests/sandbox-basic.cc:40
```

> On the child's side, the process that was blocked waiting on the one-bit semaphore is woken up, then extracts the index and call frame, and invokes the corresponding sandboxed function. 

Reviewer: ```suggestion
The pointer to the call frame object and the function index are not used again after this point and so a reentrant call (if the child calls back into the parent, which then calls back into the child) can reuse them.
```
